### PR TITLE
Refactor the codebase for hash-based artifacts versioning changes

### PR DESCRIPTION
**Pull request by remi00**
Refactor the modules identified that will have to be changed in order to address #8666 

  - Replace bluebird Promise with native one
  - Get rid of `BbPromise.bind(this)` chains
  - Add simple implementation that is the equivalent of bluebird's `Promise.map` based on this [tiny-async-pool](https://github.com/rxaviers/async-pool)

For the rollback module, there is slight behavior change: previously a string was thrown, which isn't a good practice. Replaced with an `Error`.

### CODE REVIEW

File: lib/plugins/aws/deploy/lib/uploadArtifacts.js [hunk 1]
```diff
@@ -140,7 +139,7 @@ module.exports = {
         .filter(Boolean)
     );
 
-    return BbPromise.map(
+    return promiseMap(
```
Comment (medikoo) [inline lib/plugins/aws/deploy/lib/uploadArtifacts.js:142]: Let's use `Promise.all`, and we can control concurrency as:

```javascript
const limit = require('ext/promise/limit).bind(Promise)

Promise.all(limit(MAX_CONCURRENT_ARTIFACTS_UPLOADS, async () => {...})
```

`ext` is currently not listed as our dependency, but it's used by our dependencies, so that technically doesn't add to our dependencies baggage, and it's free to add it.

File: lib/plugins/aws/package/index.js [hunk 2]
```diff
@@ -77,16 +77,17 @@ class AwsPackage {
...
+      'before:package:compileLayers': async () => {
+        await this.generateArtifactDirectoryName();
+      },
 
-      'package:finalize': async () =>
-        BbPromise.bind(this).then(() =>
-          this.serverless.pluginManager.spawn('aws:package:finalize')
-        ),
+      'package:finalize': async () => {
+        return this.serverless.pluginManager.spawn('aws:package:finalize');
```
Comment (medikoo) [inline lib/plugins/aws/package/index.js:89]: It'll be nice to have common approach, probably simplest is:

```javascript
'hook1': async () => fn(),
'hook2': async () => fn(),
```

File: lib/plugins/aws/rollback.js [hunk 3]
```diff
@@ -1,6 +1,4 @@
 'use strict';
```
Comment (medikoo) [inline lib/plugins/aws/rollback.js:1]: I don't think that we will need changes here, but anyway this refactor doesn't harm :)

File: lib/plugins/aws/rollback.js [hunk 4]
```diff
@@ -44,45 +41,41 @@ class AwsRollback {
...
 
-        if (deployments.length === 0) {
-          const msg = "Couldn't find any existing deployments.";
-          const hint = 'Please verify that stage and region are correct.';
-          return BbPromise.reject(`${msg} ${hint}`);
-        }
+    if (deployments.length === 0) {
+      const msg = "Couldn't find any existing deployments.";
+      const hint = 'Please verify that stage and region are correct.';
+      throw new Error(`${msg} ${hint}`);
```
Comment (medikoo) [inline lib/plugins/aws/rollback.js:53]: That's a good pick.

Still it feels as a _user_ error (not an error on our side), and in that case we communicate those with `ServerlessError` and we also provide a code (search for `ServerlessError` and it should bring tons of such cases)

File: lib/plugins/aws/rollback.js [hunk 5]
```diff
@@ -44,45 +41,41 @@ class AwsRollback {
...
 
-        if (!exists) {
-          const msg = `Couldn't find a deployment for the timestamp: ${this.options.timestamp}.`;
-          const hint = 'Please verify that the timestamp, stage and region are correct.';
-          return BbPromise.reject(`${msg} ${hint}`);
-        }
+    if (!exists) {
+      const msg = `Couldn't find a deployment for the timestamp: ${this.options.timestamp}.`;
+      const hint = 'Please verify that the timestamp, stage and region are correct.';
+      throw new Error(`${msg} ${hint}`);
```
Comment (medikoo) [inline lib/plugins/aws/rollback.js:75]: Same here

File: test/unit/lib/plugins/aws/rollback.test.js [hunk 6]
```diff
@@ -123,8 +123,8 @@ describe('AwsRollback', () => {
         .then(() => {
           assert.isNotOk(true, 'setStackToUpdate should not resolve');
         })
-        .catch((errorMessage) => {
-          expect(errorMessage).to.include("Couldn't find any existing deployments");
+        .catch((error) => {
+          expect(error.message).to.include("Couldn't find any existing deployments");
```
Comment (medikoo) [inline test/unit/lib/plugins/aws/rollback.test.js:127]: When switching to `ServerlessError`, we also confirm that expected error happened, just by confirming on its `code` (we do not confirm on `message` as that's subject to grammar improvements and those should not break tests)

File: test/unit/lib/plugins/aws/rollback.test.js [hunk 7]
```diff
@@ -157,8 +157,8 @@ describe('AwsRollback', () => {
         .then(() => {
           assert.isNotOk(true, 'setStackToUpdate should not resolve');
         })
-        .catch((errorMessage) => {
-          expect(errorMessage).to.include("Couldn't find a deployment for the timestamp");
+        .catch((error) => {
```
Comment (medikoo) [inline test/unit/lib/plugins/aws/rollback.test.js:160]: Same here

**Review by medikoo — CHANGES_REQUESTED** [review]
Thank you @remi00 !

It look very good in general, please see my remarks

File: lib/plugins/aws/deploy/lib/uploadArtifacts.js (hunk 1 above)
Comment (remi00) [reply lib/plugins/aws/deploy/lib/uploadArtifacts.js:142]: @medikoo Currently, latest version of [ext@1.4.0](https://www.npmjs.com/package/ext/v/1.4.0) published on NPM does not include `promise/limit` module. It is also hard to follow the versioning, as the releases on GitHub use two different versioning schemes. Could you please help with providing proper version to get promise/limit in place?

Comment (medikoo) [reply lib/plugins/aws/deploy/lib/uploadArtifacts.js:142]: @remi00 indeed! Sorry I was completely sure it's published.

It's published now with v1.5.0

**Comment by remi00** [comment]
@medikoo thanks for throrough comments, I think I addressed them all. Please take a look.

File: lib/plugins/aws/rollback.js [hunk 8]
```diff
@@ -44,45 +43,41 @@ class AwsRollback {
...
 
-        if (deployments.length === 0) {
-          const msg = "Couldn't find any existing deployments.";
-          const hint = 'Please verify that stage and region are correct.';
-          return BbPromise.reject(`${msg} ${hint}`);
-        }
+    if (deployments.length === 0) {
+      const msg = "Couldn't find any existing deployments.";
+      const hint = 'Please verify that stage and region are correct.';
+      throw new ServerlessError(`${msg} ${hint}`, 'ROLLBACK_IMPOSSIBLE_NO_DEPLOYMENTS');
```
Comment (medikoo) [inline lib/plugins/aws/rollback.js:55]: Let's maybe reword code to `ROLLBACK_DEPLOYMENTS_NOT_FOUND`

File: lib/plugins/aws/rollback.js [hunk 9]
```diff
@@ -44,45 +43,41 @@ class AwsRollback {
...
 
-        if (!exists) {
-          const msg = `Couldn't find a deployment for the timestamp: ${this.options.timestamp}.`;
-          const hint = 'Please verify that the timestamp, stage and region are correct.';
-          return BbPromise.reject(`${msg} ${hint}`);
-        }
+    if (!exists) {
+      const msg = `Couldn't find a deployment for the timestamp: ${this.options.timestamp}.`;
+      const hint = 'Please verify that the timestamp, stage and region are correct.';
+      throw new ServerlessError(`${msg} ${hint}`, 'ROLLBACK_IMPOSSIBLE_NO_DEPLOYMENT');
```
Comment (medikoo) [inline lib/plugins/aws/rollback.js:77]: Same here, maybe to `ROLLBACK_DEPLOYMENT_NOT_FOUND`

**Review by medikoo — CHANGES_REQUESTED** [review]
@remi00 looks great! I have just last few very minor remarks and we're good with it

**Review by medikoo — APPROVED** [review]
Thank you @remi00, that looks great!
